### PR TITLE
TargetID: Scalability

### PR DESCRIPTION
**Pull request by TimGoll**
image

TargetID is now scaled with the global scale factor. While I was at it, I updated some targetID variable names to have a unified naming scheme.
To make things easier in the future, I added a new library with scaled drawfunctions, named `drawsc`. Additionally I moved all remaining draw functions into their appropriate files.

### CODE REVIEW

File: lua/ttt2/extensions/draw.lua [hunk 1]
```diff
@@ -0,0 +1,436 @@
+---
+-- draw extension functions
+-- @author Mineotopia, LeBroomer, Alf21, saibotk
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: see my `surface.lua` comment

File: lua/ttt2/extensions/draw.lua [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+
+	return tmpCol
+end
+
+---
+-- A function to draw an outlined box with a definable width
+-- @param number x The x position of the rectangle
+-- @param number y The y position of the rectangle
+-- @param number w The width of the rectangle
+-- @param number h The height of the rectangle
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: box? rectangle? please choose one of the two or if you meant different things pls specify the relation of the two

File: lua/ttt2/extensions/draw.lua [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+end
+
+local drawOutlinedBox = draw.OutlinedBox
+
+---
+-- A function to draw an outlined box with a shadow
+-- @param number x The x position of the rectangle
+-- @param number y The y position of the rectangle
+-- @param number w The width of the rectangle
+-- @param number h The height of the rectangle
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: same as above

File: lua/ttt2/extensions/draw.lua [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+-- @param [default=1]number t The thickness of the line
+-- @param [default=Color(255,255,255,255)]Color color The color of the line
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: `line` -> `outline` maybe?

File: lua/ttt2/extensions/draw.lua [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+	local tmpCol = GetShadowColor(color)
+
+	drawOutlinedBox(x + shift2, y + shift2, w, h, t, tmpCol)
+	drawOutlinedBox(x + shift1, y + shift1, w, h, t, tmpCol)
+	drawOutlinedBox(x + shift1, y + shift1, w, h, t, tmpCol)
+	drawOutlinedBox(x, y, w, h, t, color)
+end
+
+---
+-- A function to draws a simple box without a radius
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: why is it needed to specify "without a radius" what is the "radius" you speak of?

File: lua/ttt2/extensions/draw.lua [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+-- @realm client
+function draw.Box(x, y, w, h, color)
+	surface.SetDrawColor(color or COLOR_WHITE)
+	surface.DrawRect(x, y, w, h)
+end
+
+local drawBox = draw.Box
+
+---
+-- A function to draws a simple shadowed box without a radius
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: same as above

File: lua/ttt2/extensions/draw.lua [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+	drawOutlinedCircle(x, y, r, color)
+end
+
+---
+-- A function to draw an outlined box with a shadow
+-- @param number startX The x position to start the line
+-- @param number startY The y position to start the line
+-- @param number endX The x position to end the line
+-- @param number endY The y position to end the line
+-- @param [default=Color(255,255,255,255)]Color color The color of the line
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: `line` -> `outline` maybe?

File: lua/ttt2/extensions/draw.lua [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+
+local drawLine = draw.Line
+
+---
+-- A function to draw an outlined box with a shadow
+-- @param number startX The x position to start the line
+-- @param number startY The y position to start the line
+-- @param number endX The x position to end the line
+-- @param number endY The y position to end the line
+-- @param [default=Color(255,255,255,255)]Color color The color of the line
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: same as above

File: lua/ttt2/extensions/draw.lua [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+end
+
+---
+-- Draws a filtered textured rectangle / image / icon
+-- @param number x
+-- @param number y
+-- @param number w width
+-- @param number h height
+-- @param Material material
+-- @param [default=255]number alpha
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: missing param descriptions

File: lua/ttt2/extensions/draw.lua [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+local drawFilteredTexture = draw.FilteredTexture
+
+---
+-- Draws a filtered textured rectangle / image / icon with shadow
+-- @param number x
+-- @param number y
+-- @param number w width
+-- @param number h height
+-- @param Material material
+-- @param [default=255]number alpha
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: missing param descriptions

File: lua/ttt2/extensions/draw.lua [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+	drawSimpleText(text, font, x + shift2, y + shift2, tmpCol, xalign, yalign)
+	drawSimpleText(text, font, x + shift1, y + shift1, tmpCol, xalign, yalign)
+	drawSimpleText(text, font, x + shift1, y + shift1, tmpCol, xalign, yalign)
+	drawSimpleText(text, font, x, y, color, xalign, yalign)
+end
+
+local drawShadowedText = draw.ShadowedText
+
+---
+-- Draws an advanced text (scalable)
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: scalable by what?

File: lua/ttt2/extensions/draw.lua [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+-- @note You should use @{surface.CreateAdvancedFont} before trying to access the font
+-- @2D
+-- @param string text The text to be drawn
+-- @param [default="DefaultBold"] string font The font. See @{surface.CreateAdvancedFont} to create your own. The original font should be always created, see @{surface.CreateFont}.
+-- @param number x The X Coordinate
+-- @param number y The Y Coordinate
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: The coordinate of what? left upper corner? center?

File: lua/ttt2/extensions/draw.lua [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+-- @param Color color The color of the text. Uses the Color structure.
+-- @param number xalign The alignment of the X coordinate using
+-- <a href="https://wiki.garrysmod.com/page/Enums/TEXT_ALIGN">TEXT_ALIGN_Enums</a>.
+-- @param number yalign The alignment of the Y coordinate using
+-- <a href="https://wiki.garrysmod.com/page/Enums/TEXT_ALIGN">TEXT_ALIGN_Enums</a>.
+-- @param boolean shadow whether there should be a shadow of the text
+-- @param number scale scale (float number)
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: unfinished

File: lua/ttt2/extensions/draw.lua [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+end
+
+---
+-- Returns a list of lines to wrap the text matching the given width
+-- @param string text
+-- @param number width
+-- @param string font
+-- @param number scale
+-- @return table
+-- @realm client
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: unfinished doc comment

File: lua/ttt2/extensions/draw.lua [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+function draw.GetWrappedText(text, width, font, scale)
+	-- Oh joy, I get to write my own wrapping function. Thanks Lua!
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: why is this here?

File: lua/ttt2/extensions/draw.lua [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+
+		if line_w > length then
+			length = line_w
+		end
+	end
+
+	-- get height of lines
+	local _, line_h = surface.GetTextSize(text)
+
+	return lines, length * scale, line_h * lns * scale
```
Comment (Histalek) [inline lua/ttt2/extensions/draw.lua]: what are these random comments? please remove them or write them a way they are actually useful.
in-code comments should not explain what is done but rather why it is done

File: gamemodes/terrortown/gamemode/shared/sh_init.lua [hunk 17]
```diff
@@ -523,9 +523,17 @@ require("marks")
...
 include("ttt2/extensions/net.lua")
 include("ttt2/extensions/string.lua")
 include("ttt2/extensions/table.lua")
+include("ttt2/extensions/surface.lua")
+include("ttt2/extensions/draw.lua")
+
+-- include libraries
+include("ttt2/libraries/fonts.lua")
+include("ttt2/libraries/drawsc.lua")
 include("ttt2/extensions/draw.lua")
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/shared/sh_init.lua:537]:
```suggestion
include("ttt2/extensions/draw.lua")

-- include libraries
include("ttt2/libraries/fonts.lua")
include("ttt2/libraries/drawsc.lua")
```

File: gamemodes/terrortown/gamemode/shared/sh_init.lua [hunk 18]
```diff
@@ -523,9 +523,17 @@ require("marks")
...
 
 include("ttt2/libraries/appearance.lua")
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/shared/sh_init.lua:539]:
```suggestion
include("ttt2/extensions/draw.lua")

-- include libraries
include("ttt2/libraries/fonts.lua")
include("ttt2/libraries/drawsc.lua")
include("ttt2/libraries/appearance.lua")
```

File: CHANGELOG.md [hunk 19]
```diff
@@ -10,12 +10,15 @@ All notable changes to TTT2 will be documented here. Inspired by [keep a changel
   - introduced new development interfaces to easily create menues and settings for addons
 - Introduced a global scale factor based on screen resolution to scale HUD elements accordingly
 - Added automatical scale factor change on resolution change that works even if the resolution was changed while TTT2 wasn't loaded
+- Added scale based draw functions
```
Comment (Histalek) [inline CHANGELOG.md:13]: i think it should be pointed out that this is a library, so it can be found more easily in the codebase.
something along the lines of:
```suggestion
- Added `drawsc` library featuring scalable draw functions
```

File: lua/ttt2/extensions/draw.lua [hunk 20]
```diff
@@ -53,7 +64,7 @@ local drawOutlinedBox = draw.OutlinedBox
 function draw.OutlinedShadowedBox(x, y, w, h, t, color)
 	color = color or COLOR_WHITE
 
-	local tmpCol = color.r + color.g + color.b > 200 and tableCopy(shadowColorDark) or tableCopy(shadowColorWhite)
+	local tmpCol = color.r + color.g + color.b > 200 and tableCopy(colorShadowDark) or tableCopy(colorShadowBright)
```
Comment (Alf21) [inline lua/ttt2/extensions/draw.lua:67]: You should use the function GetShadowColor()

File: lua/ttt2/libraries/drawsc.lua [hunk 21]
```diff
@@ -0,0 +1,220 @@
+---
+-- drawsc library functions
+-- adds a secondary function that draws based on the global scale factor
+-- which is based on a normal 1080p resolution
```
Comment (Histalek) [inline lua/ttt2/libraries/drawsc.lua:4]:
```suggestion
-- which is based on a 1920x1080p resolution
```

File: lua/ttt2/extensions/draw.lua [hunk 22]
```diff
@@ -76,8 +87,35 @@ function draw.Box(x, y, w, h, color)
...
+local drawBox = draw.Box
+
 ---
--- A function to draw a circle outline
+-- A function to draws a simple shadowed rectilinear box.
+-- @param number x The x position to start the box
+-- @param number y The y position to start the box
+-- @param number w The width of the box
+-- @param number h The height of the box
+-- @param [default=Color(255,255,255,255)]Color color The color of the box
```
Comment (Alf21) [inline lua/ttt2/extensions/draw.lua:98]:
```suggestion
-- @param Color[default=Color(255,255,255,255)] color The color of the box
```

File: lua/ttt2/extensions/draw.lua [hunk 23]
```diff
@@ -76,8 +87,35 @@ function draw.Box(x, y, w, h, color)
...
+-- @param [default=1.0]number scale A scaling factor that is used for the shadows
```
Comment (Alf21) [inline lua/ttt2/extensions/draw.lua:99]:
```suggestion
-- @param number[default=1.0] scale A scaling factor that is used for the shadows
```

Comment (Alf21) [reply lua/ttt2/extensions/draw.lua:99]: Same below

File: lua/ttt2/extensions/draw.lua [hunk 24]
```diff
@@ -118,7 +182,7 @@ local drawLine = draw.Line
 function draw.ShadowedLine(startX, startY, endX, endY, color)
 	color = color or COLOR_WHITE
 
-	local tmpCol = color.r + color.g + color.b > 200 and tableCopy(shadowColorDark) or tableCopy(shadowColorWhite)
+	local tmpCol = color.r + color.g + color.b > 200 and tableCopy(colorShadowDark) or tableCopy(colorShadowBright)
```
Comment (Alf21) [inline lua/ttt2/extensions/draw.lua:185]: Use GetShadowColor()

File: lua/ttt2/extensions/draw.lua [hunk 25]
```diff
@@ -173,7 +237,7 @@ function draw.FilteredShadowedTexture(x, y, w, h, material, alpha, color, scale)
 	color = color or COLOR_WHITE
 	scale = scale or 1
 
-	local tmpCol = color.r + color.g + color.b > 200 and tableCopy(shadowColorDark) or tableCopy(shadowColorWhite)
+	local tmpCol = color.r + color.g + color.b > 200 and tableCopy(colorShadowDark) or tableCopy(colorShadowBright)
```
Comment (Alf21) [inline lua/ttt2/extensions/draw.lua:240]: GetShadowColor()

**Review by Alf21 — CHANGES_REQUESTED** [review]

**Review by Histalek — COMMENTED** [review]
two small doc nitpicks
otherwise LGTM

**Review by Histalek — APPROVED** [review]
LGTM

File: lua/ttt2/libraries/fonts.lua [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+local isvector = isvector
+local tonumber = tonumber
+local tostring = tostring
+
+fonts = {}
+fonts.fonts = {}
+fonts.scales = {1, 1.5, 2, 2.5}
+
+---
+-- Gets the scale modifer based on a given scale.
```
Comment (saibotk) [inline lua/ttt2/libraries/fonts.lua:21]: This should be more expressive, what is a scale modifier based on a given scale? What does this function even provide? Which use case is associated with it?

File: lua/ttt2/libraries/fonts.lua [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+-- @param number scale The font scale
```
Comment (saibotk) [inline lua/ttt2/libraries/fonts.lua:22]: Is this really just a number? In the code you also allow a vector.

**Review by saibotk — CHANGES_REQUESTED** [review]

File: lua/ttt2/libraries/fonts.lua [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+
+		fonts.fonts[name][scale] = nameScaled
+	end
+end
+
+---
+-- Returns a specified font.
+-- @param string name The font name
+-- @return table A table of the font with different scales
+-- @reaöm client
```
Comment (saibotk) [inline lua/ttt2/libraries/fonts.lua:71]: `realm`

**Review by saibotk — APPROVED** [review]
LGTM

**Review by Alf21 — APPROVED** [review]
